src/passes/OptimizeInstructions.cpp

```diff
@@ -460,6 +460,101 @@ struct OptimizeInstructions
           return c;
         }
       }
+      {
+        Const* c;
+        Binary* bin;
+        Expression *x, *y;
+        BinaryOp op;
+
+        // fneg(x) * fneg(y)   ==>   x * y
+        // fneg(x) / fneg(y)   ==>   x / y
+        if (matches(
+              curr,
+              binary(&bin, Mul, unary(Neg, any(&x)), unary(Neg, any(&y)))) ||
+            matches(
+              curr,
+              binary(&bin, DivS, unary(Neg, any(&x)), unary(Neg, any(&y))))) {
+          bin->left = x;
+          bin->right = y;
+          return bin;
+        }
+        // fneg(x * C)   ==>   x * -C
+        // fneg(x / C)   ==>   x / -C
+        // fneg(C / x)   ==>   -C / x
+        //
+        // TODO: using `fastMath` is not necessary, but due to fuzzer can't
+        // flexible handle sign of NaNs we use it here.
+        if (fastMath &&
+            (matches(curr, unary(Neg, binary(&bin, Mul, any(&x), fval(&c)))) ||
+             matches(curr, unary(Neg, binary(&bin, DivS, any(&x), fval(&c)))) ||
+             matches(curr,
+                     unary(Neg, binary(&bin, DivS, fval(&c), any(&x)))))) {
+          c->value = c->value.neg();
+          return bin;
+        }
+        // (fneg(x) * y) op C   ==>   (x * y) op -C
+        // (x * fneg(y)) op C   ==>   (x * y) op -C
+        // (fneg(x) / y) op C   ==>   (x / y) op -C
+        // (x / fneg(y)) op C   ==>   (x / y) op -C
+        // (x * fneg(y))  + C   ==>   C - (x * y)
+        // (x / fneg(y))  + C   ==>   C - (x / y)
+        //
+        // where op = `*` or `/`
+        if ((matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, any(&x), unary(Neg, any(&y))),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, any(&x), unary(Neg, any(&y))),
+                            fval(&c)))) &&
+            (op == Abstract::getBinary(bin->type, Mul) ||
+             op == Abstract::getBinary(bin->type, DivS) ||
+             (fastMath && op == Abstract::getBinary(bin->type, Add)))) {
```

Comment:
fastMath should not be needed here. The final operation is not an fneg, on either side - so the interpreter will emit a canonical nan if the result is a nan. (The only risk is if the final operation is an fneg on just one side - the fneg does *not* canonicalize the nan, so it would flip the sign, which would look different.)

Also, please move the addition case out of this - I think that would be simpler? As it flips the output, and changes + to -, it doesn't really share that much with the * / cases.